draft-ietf-mls-protocol.md

```diff
@@ -494,26 +502,105 @@ as follows:
-primitives, defined by the ciphersuite in use:
+A particular instance of a ratchet tree is based on the following
+cryptographic primitives, defined by the ciphersuite in use:
 
 * A Diffie-Hellman finite-field group or elliptic curve
 * A Key Derivation Function (KDF)
 * A Derive-Key-Pair function that produces an asymmetric keypair
   from a node secret
 
-A ratchet tree is a left-balanced binary tree, in which each node
-contains up to three values:
+Each node in a ratchet tree contains up to three values:
+
+* A private key (only within direct path, see below)
+* A public key
+* A credential (only for leaf nodes)
+
+The conditions under which each of these values must or must not be
+present are laid out in {{views}}.
+
+A node in the tree may also be _blank_, indicating that no value is
+present at that node.  The _resolution_ of a node is an ordered list
+of non-blank nodes that collectively cover all non-blank descendants
+of the node.  The nodes in a resolution are ordered according to
+their indices.
+
+* The resolution of a non-blank node is a one element list
+  containing the node itself
+* The resolution of a blank leaf node is the empty list
+* The resolution of a blank intermediate node is the result of
+  concatinating the resolution of its left child with the resolution
+  of its right child, in that order
+
+For example, consider the following tree, where the "\_" character
+represents a blank node:
+
+~~~~~
+      _
+    /   \
+   /     \
+  _       CD
+ / \     / \
+A   _   C   D
+
+0 1 2 3 4 5 6
+~~~~~
+
+In this tree, we can see all three of the above rules in play:
+
+* The resolution of node 5 is the list [CD]
+* The resolution of node 2 is the empty list []
+* The resolution of node 3 is the list [A, CD]
+
+Every node, regardless of whether a node is blank or populated, has
+a corresponding _hash_ that summarizes the contents of the subtree
+below that node.  The rules for computing these hashes are described
+in {{tree-hashes}}.
+
+## Views of a Ratchet Tree {#views}
+
+MLS assumes that each participant maintains a complete and
+up-to-date view of the public state of the group's ratchet tree,
+including the public keys for all nodes and the credentials
+associated with the leaf nodes.
 
-* A secret octet string (optional)
-* An asymmetric private key (optional)
-* An asymmetric public key
+No participant in an MLS group has full knowledge of the secret
+state of the tree, i.e., the secret values and private keys associated to
+the nodes.  Instead, each member is assigned to a leaf of the tree,
+which determines the set of secret state known to the member.  The
+credential stored at that leaf is one provided by the member.
+
+In particular, MLS maintains the members view of the tree in such a
+way as to maintain the _tree invariant_:
+
+    The secret value and private key for a node in the tree are
+    known to a member of the group if and only if that member's leaf
+    is a descendent of the node or equal to it.
```

Comment:
Typo (unless you really mean it): "descendent" -> "descendant". The latter is how it's spelled everywhere else in the doc